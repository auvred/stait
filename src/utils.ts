export type ContextDecl = Record<string, unknown>

export type ActionsDecl<_Typestait extends { type: string }> = Record<
  string,
  // `any` looks cleaner in user-facing errors
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (payload?: any) => _Typestait['type']
>

export type Flatten<T> = [T] extends [infer TT]
  ? {
      [K in keyof TT]: TT[K]
    }
  : never
