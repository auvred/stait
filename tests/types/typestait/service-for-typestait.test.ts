import { runTypesTests } from '../../utils'

import type { ServiceForTypestait } from '../../../src'
import type { Equal } from '../../utils'

runTypesTests(import.meta.url)

type ExpectType_recursive_service_depth_1 = ServiceForTypestait<
  {
    type: 'stait1'
    service: 'stait1'
  },
  'stait1'
>

type ExpectType_recursive_service_depth_2 = Awaited<
  ReturnType<ExpectType_recursive_service_depth_1>
>['service']

type ExpectType_recursive_service_depth_2_equals_depth_1 = Equal<
  ExpectType_recursive_service_depth_1,
  ExpectType_recursive_service_depth_2
>

type ExpectType_multiple_staits = ServiceForTypestait<
  | {
      type: 'stait1'
      service: 'stait2'
    }
  | {
      type: 'stait2'
      context: {
        data1: string
        data2: 123
      }
    },
  'stait1'
>

type ExpectError_service_for_non_existent_stait = ServiceForTypestait<
  {
    type: 'stait1'
    service: 'stait1'
  },
  'stait2'
>
