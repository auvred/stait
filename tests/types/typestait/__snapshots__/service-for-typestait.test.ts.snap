// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`error: service for non existent stait 1`] = `
"error TS2344: Type '\\"stait2\\"' does not satisfy the constraint '\\"stait1\\"'.
"
`;

exports[`type: multiple staits 1`] = `
"type multiple_staits =
  (
    context: unknown,
  ) => MaybePromise<{
    type: 'stait2'
    context: {
      data1: string
      data2: 123
    }
  }>
"
`;

exports[`type: recursive service depth 1 1`] = `
"type recursive_service_depth_1 =
  (
    context: unknown,
  ) => MaybePromise<{
    type: 'stait1'
    service: any
  }>
"
`;

exports[`type: recursive service depth 2 1`] = `
"type recursive_service_depth_2 =
  (
    context: unknown,
  ) => MaybePromise<{
    type: 'stait1'
    service: any
  }>
"
`;

exports[`type: recursive service depth 2 equals depth 1 1`] = `
"type recursive_service_depth_2_equals_depth_1 =
  true
"
`;
